### DISCUSSION TITLE
Acceptance Rate

### DISCUSSION THREAD

**Discussion by Peku995** [cdslaborg/paramonte#26] [Q&A] [answered]
Hi,

I was wondering what is a suitable acceptance rate for the sampler?
I read online for metropolis hastings it should be between 10% to 60%, but is there a recommendation for Paramonte?

Thanks

Peter

**Comment by shahmoradi** [comment, marked as answer]
@Peku995 a reasonable acceptance rate often depends on the specific problem being solved. 
60% is likely too high for any target function with more than 1 dimension. 
The theoretical optimal acceptance rate is around `0.234` (or `23.4%`) for simple infinite-dimensional Multivariate Normal distribution as the target density.
But in practice, it is usually lower for medium dimensional problems (5-50 dimensions, perhaps more around 10-20%).
The paradram sampler has a [targetAcceptanceRate](https://www.cdslab.org/paramonte/notes/usage/paradram/specifications/#targetacceptancerate) specification that sets a bracket for the allowed range of acceptance rate. By default, it is `[0, 1]`, meaning that the sampler will approach a value in this range dictated by the problem and initial conditions. You can change this to narrower ranges if you like, but I recommend using it only during the exploration phase. 
In sum, I recommend setting the initial conditions such that the acceptance rate approaches a mean value of around 10-20% for medium-dimensional problems. As the number of dimensions increases, you may want to reduce this mean rate by adjusting the initial conditions, such as increasing the [proposal scale factor](https://www.cdslab.org/paramonte/notes/usage/paradram/specifications/#scalefactor).
If you need further help, I'd be happy to help. 
@ShashankKumbhare is currently working on a new pending release of the sampler, which fully automates the adjustment of the acceptance rate on the fly so that it remains near optical for arbitrary problems in arbitrary dimensions.

**Reply by Peku995** [reply]
Thank you for your answer! 
This leads to a follow-up question: You wrote in the second part of your answer that the proposal scale factor should be set in order to achieve the desired acceptance rate. What is the difference if I set the targetAcceptranceRate directly? Or is there no difference?

An automatic adjustment of the acceptrance rate sounds great!

**Reply by shahmoradi** [reply]
@Peku995 That's an excellent question:
- Scaling the proposal by an initial [scale factor](https://www.cdslab.org/paramonte/notes/usage/paradram/specifications/#scalefactor) shifts the sampler's acceptance rate to higher or lower values globally throughout the simulation. It has the benefit of retaining the Markovian property of the final chain. Still, it comes with the caveat of being less flexible and adaptive in response to phase transitions in the target function (e.g., discovering new modes or new landscapes that look significantly different from what has been explored in the past.
- By contrast, narrowing the acceptance rate by [targetAcceptanceRate](https://www.cdslab.org/paramonte/notes/usage/paradram/specifications/#targetacceptancerate) yields a flexible sampler that can respond to runtime surprises (like phase transitions) very well and by adapting the proposal scale to keep the acceptance rate in the requested range on the fly. But in doing so, it will break the Markovian property of the resulting chain, meaning that the generated sample will not have necessarily come from the specified target density (e.g., the final sample may not be [iid](https://en.wikipedia.org/wiki/Independent_and_identically_distributed_random_variables). 
To address this potential weakness, the sampler also generates an output variable in the `*_chain.txt` file named `AdapatationMeasure`. Under normal conditions, this variable should steadily decline from `1` to `0` throughout the chain (see the plot below).
![image](https://github.com/cdslaborg/paramonte/assets/4683401/720671d2-623b-4a3c-a0f1-6608adde408c)

This shape of the curve in this example implies that initially, the proposal is strongly changing its shape and scale to adapt to the shape of the target function, but as it learns the target function, the sampler gradually stops adapting the shape and scale of the proposal. This is the ideal behavior to see in the simulation output. If this curve takes an upward trend at any simulation stage, it implies that the sampler has likely entered a new phase in the target function (e.g., possibly identified a new higher function peak that looks different from the previous). Ultimately, this variable should decline to values below `0.001` to safely assume that the simulation has converged and the final mcmc chain is practically Markovian. You can find more theoretical details of this approach near Figure 8 in [shahmoradi and Bagheri (2020)](https://www.cdslab.org/paramonte/notes/overview/preface/#the-paradram-sampler).
This [matlab live notebook](https://www.cdslab.org/paramonte/notes/examples/matlab/mlx/sampling_multivariate_normal_distribution_via_paradram.html) and [Python Jupyter notebook](https://nbviewer.org/github/cdslaborg/paramontex/blob/main/Python/Jupyter/sampling_multivariate_normal_distribution_via_paradram/sampling_multivariate_normal_distribution_via_paradram.ipynb) contain scripts to visualize the output `AdaptationMeasure`.

**Reply by Peku995** [reply]
Thank you very much for your answer!

**Comment by arwelHughes** [comment]
Also I second that - thank you for your answer Amir!

Personally, I've found that 'tuning' the scale factor makes a huge difference to the speed of the simulation for our class of problems. The result is the same eventually, but the choosing the scalefcator correctly can be the difference between running the same simulation in either 1 minute or 10 minutes. Automatically choosing this sounds like a great idea.....
________________________________
